File: contribution-pipeline/scheduler/backend/src/idempotency.ts
/**
 * Idempotency-Key 支持 (含 M5 严格并发去重)
 *
 * 用法:
 *   1) 调用方在 POST 请求加 header `Idempotency-Key: <opaque-string>` (建议 UUID)
 *   2) preHandler 拿 PG advisory_xact_lock + 占位:
 *      - 第一次 → INSERT status='pending', 继续 handler (onSend 收尾 → status='completed')
 *      - 同 key + 同 hash + completed → 重放原响应
 *      - 同 key + 同 hash + pending → 409 IDEMPOTENCY_IN_PROGRESS (并发, 客户端串行重试)
 *      - 同 key + 不同 hash → 409 IDEMPOTENCY_HASH_MISMATCH (误用)
 *   3) 缓存 24h 后过期; pending 行 5 分钟未完成视为 stuck, 后续请求可覆盖
 *
 * 适用接口: 创建型/状态推进型 POST (`/items/create`, `/result`, `/api/batches`).
 * 不适用: 纯查询 GET, 或本身用 runId 做天然幂等的内部接口.
 */

import { createHash } from "node:crypto";
import type { FastifyInstance, FastifyRequest } from "fastify";
import { asSystem } from "./db.ts";
import { DEFAULT_TENANT_ID } from "./auth.ts";

export const IDEMPOTENCY_HEADER = "idempotency-key";

const PENDING_TTL_SEC = 5 * 60;       // pending 行有效 5 分钟
const COMPLETED_TTL_SEC = 24 * 3600;  // completed 行缓存 24 小时

interface ExistingRow {
  request_hash: string;
  status: "pending" | "completed";
  status_code: number | null;
  response_body: unknown;
}

function bodyHash(body: unknown): string {
  const s = body == null ? "" : JSON.stringify(body);
  return createHash("sha256").update(s).digest("hex");
}

/**
 * 路由级注册: 把要做幂等的路由路径加进 `protectedRoutes` 集合,
 * preHandler 钩子拿锁占位, onSend 钩子收尾.
 */
export function registerIdempotency(
  app: FastifyInstance,
  protectedRoutes: { method: string; url: string }[],
) {
  const protectedSet = new Set(
    protectedRoutes.map((r) => `${r.method.toUpperCase()} ${r.url}`),
  );

  app.decorateRequest("idemKey", null as null | string);
  app.decorateRequest("idemScope", null as null | string);
  app.decorateRequest("idemHash", null as null | string);
  app.decorateRequest("idemTenant", null as null | string);
  app.decorateRequest("idemReplayed", false);
  app.decorateRequest("idemNeedsCompletion", false);  // M5: 是否要在 onSend 收尾

  app.addHook("preHandler", async (req, reply) => {
    const routeKey = `${req.method.toUpperCase()} ${req.routeOptions.url ?? req.url}`;
    if (!protectedSet.has(routeKey)) return;

    const key = req.headers[IDEMPOTENCY_HEADER];
    if (!key || typeof key !== "string") return;
    if (key.length < 8 || key.length > 200) {
      return reply.code(400).send({
        error: { code: "BAD_IDEMPOTENCY_KEY", message: "Idempotency-Key 长度必须在 8-200 之间" },
      });
    }

    const scope = req.routeOptions.url ?? req.url;
    const hash = bodyHash(req.body);
    const tenantId = req.caller?.tenantId ?? DEFAULT_TENANT_ID;
    (req as any).idemKey = key;
    (req as any).idemScope = scope;
    (req as any).idemHash = hash;
    (req as any).idemTenant = tenantId;

    // M5: 单事务内拿 advisory lock + 查现状 + 占位. lock 在 COMMIT 时自动释放,
    // 同 key 并发请求会串行排队 (第二个看到 pending 行 → IDEMPOTENCY_IN_PROGRESS).
    type Decision =
      | { kind: "first" }
      | { kind: "replay"; status_code: number; response_body: unknown }
      | { kind: "in_progress" }
      | { kind: "hash_mismatch" };

    const decision: Decision = await asSystem(async (tx) => {
      // hashtextextended 同一 (tenant, key, scope) 必同, 不同 key 极小概率撞 → 锁串行有限可接受
      const lockKey = createHash("sha256")
        .update(`${tenantId}|${key}|${scope}`)
        .digest()
        .readBigInt64BE(0);
      // postgres.js 不直接接受 BigInt 给 ?, 用字符串
      await tx`SELECT pg_advisory_xact_lock(${lockKey.toString()}::bigint)`;

      // 抽样清过期 (1%)
      if (Math.random() < 0.01) {
        await tx`DELETE FROM idempotency_keys WHERE expires_at < NOW()`.catch(() => {});
      }

      const existingRows = await tx<ExistingRow[]>`
        SELECT request_hash, status, status_code, response_body
        FROM idempotency_keys
        WHERE tenant_id = ${tenantId} AND key = ${key} AND scope = ${scope}
          AND expires_at > NOW()
      `;
      const existing = existingRows[0];

      if (!existing) {
        // 没记录, INSERT pending 占位 (UPSERT 兼顾"老 stuck 行已过期"被 SELECT 漏掉的边界)
        await tx`
          INSERT INTO idempotency_keys
            (tenant_id, key, scope, request_hash, status, status_code, response_body, expires_at)
          VALUES
            (${tenantId}, ${key}, ${scope}, ${hash}, 'pending', NULL, NULL,
             NOW() + (${PENDING_TTL_SEC} || ' seconds')::interval)
          ON CONFLICT (tenant_id, key, scope) DO UPDATE SET
            request_hash = EXCLUDED.request_hash,
            status = 'pending',
            status_code = NULL,
            response_body = NULL,
            expires_at = EXCLUDED.expires_at,
            created_at = NOW()
        `;
        return { kind: "first" };
      }

      if (existing.request_hash !== hash) {
        return { kind: "hash_mismatch" };
      }
      if (existing.status === "completed") {
        return {
          kind: "replay",
          status_code: existing.status_code ?? 200,
          response_body: existing.response_body,
        };
      }
      // status='pending' + hash 一致: 同 key 并发. 让客户端串行重试 (而不是穿透).
      return { kind: "in_progress" };
    });

    if (decision.kind === "first") {
      (req as any).idemNeedsCompletion = true;
      return;  // 继续 handler
    }
    if (decision.kind === "replay") {
      (req as any).idemReplayed = true;
      return reply.code(decision.status_code).send(decision.response_body);
    }
    if (decision.kind === "in_progress") {
      return reply.code(409).send({
        error: {
          code: "IDEMPOTENCY_IN_PROGRESS",
          message: "同 Idempotency-Key 的请求正在处理, 请稍后重试",
        },
      });
    }
    return reply.code(409).send({
      error: {
        code: "IDEMPOTENCY_HASH_MISMATCH",
        message: "相同 Idempotency-Key 已用于不同请求体",
      },
    });
  });

  // onSend: 把首次响应缓存下来 (4xx/5xx 也存, 避免错误重试穿透)
  app.addHook("onSend", async (req, reply, payload) => {
    if (!(req as any).idemNeedsCompletion) return payload;
    const key = (req as any).idemKey as string;
    const scope = (req as any).idemScope as string;
    const hash = (req as any).idemHash as string;
    const tenantId = (req as any).idemTenant as string;

    let body: unknown = payload;
    if (typeof payload === "string") {
      try { body = JSON.parse(payload); } catch { /* 留作字符串 */ }
    }
    try {
      await asSystem(async (tx) => {
        await tx`
          UPDATE idempotency_keys SET
            status = 'completed',
            status_code = ${reply.statusCode},
            response_body = ${tx.json(body as any)},
            request_hash = ${hash},
            expires_at = NOW() + (${COMPLETED_TTL_SEC} || ' seconds')::interval
          WHERE tenant_id = ${tenantId} AND key = ${key} AND scope = ${scope}
        `;
      });
    } catch (e) {
      req.log.warn({ err: e }, "idempotency completion write failed");
    }
    return payload;
  });
}

File: contribution-pipeline/scheduler/backend-py/src/scheduler/idempotency.py
"""
Idempotency-Key 中间件 — 与 idempotency.ts 1:1, 含 M5 严格并发去重.

  - 路由白名单 (server.py 注入). 命中且带 idempotency-key header 才介入.
  - 单事务内拿 pg_advisory_xact_lock((tenant_id|key|scope)) 串行同 key 并发请求.
  - 状态机:
      first        → INSERT pending, 继续 handler, onSend 升级 completed
      replay       → 直接回放原响应
      in_progress  → 409 IDEMPOTENCY_IN_PROGRESS
      hash_mismatch→ 409 IDEMPOTENCY_HASH_MISMATCH
  - TTL: pending 5 min, completed 24 h.
  - DB scheme: idempotency_keys(tenant_id, key, scope, request_hash, status, status_code, response_body, expires_at).
"""
from __future__ import annotations
import hashlib
import json
from typing import Any

from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.responses import JSONResponse

from .auth import DEFAULT_TENANT_ID
from .db import as_system

IDEMPOTENCY_HEADER = "idempotency-key"
PENDING_TTL_SEC = 5 * 60
COMPLETED_TTL_SEC = 24 * 3600


def _body_hash(raw: bytes) -> str:
    return hashlib.sha256(raw if raw else b"").hexdigest()


def _lock_key(tenant_id: str, key: str, scope: str) -> int:
    digest = hashlib.sha256(f"{tenant_id}|{key}|{scope}".encode()).digest()
    return int.from_bytes(digest[:8], "big", signed=True)


def _err(code: str, message: str, status: int) -> JSONResponse:
    return JSONResponse(status_code=status, content={"error": {"code": code, "message": message}})


class IdempotencyMiddleware(BaseHTTPMiddleware):
    def __init__(self, app, routes: list[dict[str, str]]) -> None:
        super().__init__(app)
        self._routes: set[tuple[str, str]] = {
            (r["method"].upper(), r["path"]) for r in routes
        }

    def _matches(self, method: str, path: str) -> bool:
        return (method.upper(), path) in self._routes

    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        if not self._matches(request.method, request.url.path):
            return await call_next(request)

        idem_key = request.headers.get(IDEMPOTENCY_HEADER)
        if not idem_key:
            return await call_next(request)
        if not (8 <= len(idem_key) <= 200):
            return _err("BAD_IDEMPOTENCY_KEY",
                        "Idempotency-Key 长度必须在 8-200 之间", 400)

        caller = getattr(request.state, "caller", None)
        tenant_id = caller.tenant_id if caller else DEFAULT_TENANT_ID
        scope = request.url.path
        body_bytes = await request.body()
        hash_ = _body_hash(body_bytes)
        lock_id = _lock_key(tenant_id, idem_key, scope)

        # 让下游 handler 重新读 body (Starlette 默认 body 只能消费一次)
        async def _receive():
            return {"type": "http.request", "body": body_bytes, "more_body": False}
        request._receive = _receive  # type: ignore[attr-defined]

        decision_kind: str = ""
        replay_status: int = 200
        replay_body: Any = None

        async def _decide(conn) -> None:
            nonlocal decision_kind, replay_status, replay_body
            await conn.execute("SELECT pg_advisory_xact_lock($1)", lock_id)
            existing = await conn.fetchrow(
                """SELECT request_hash, status, status_code, response_body
                     FROM idempotency_keys
                    WHERE tenant_id = $1::uuid AND key = $2 AND scope = $3
                      AND expires_at > NOW()""",
                tenant_id, idem_key, scope,
            )
            if existing is None:
                await conn.execute(
                    """INSERT INTO idempotency_keys
                         (tenant_id, key, scope, request_hash, status, expires_at)
                       VALUES ($1::uuid, $2, $3, $4, 'pending',
                               NOW() + ($5 || ' seconds')::interval)
                       ON CONFLICT (tenant_id, key, scope) DO UPDATE SET
                         request_hash = EXCLUDED.request_hash,
                         status        = 'pending',
                         status_code   = NULL,
                         response_body = NULL,
                         expires_at    = EXCLUDED.expires_at,
                         created_at    = NOW()""",
                    tenant_id, idem_key, scope, hash_, str(PENDING_TTL_SEC),
                )
                decision_kind = "first"
                return
            if existing["request_hash"] != hash_:
                decision_kind = "hash_mismatch"
                return
            if existing["status"] == "completed":
                decision_kind = "replay"
                replay_status = existing["status_code"] or 200
                replay_body = existing["response_body"]
                return
            decision_kind = "in_progress"

        await as_system(_decide)

        if decision_kind == "replay":
            return JSONResponse(content=replay_body, status_code=replay_status)
        if decision_kind == "in_progress":
            return _err("IDEMPOTENCY_IN_PROGRESS",
                        "同 Idempotency-Key 的请求正在处理, 请稍后重试", 409)
        if decision_kind == "hash_mismatch":
            return _err("IDEMPOTENCY_HASH_MISMATCH",
                        "相同 Idempotency-Key 已用于不同请求体", 409)

        # first → 跑 handler + 收尾
        response = await call_next(request)
        captured = b""
        async for chunk in response.body_iterator:
            captured += chunk if isinstance(chunk, bytes) else chunk.encode()

        try:
            parsed_body: Any = json.loads(captured) if captured else None
        except Exception:
            parsed_body = captured.decode(errors="replace") if captured else None

        async def _complete(conn) -> None:
            await conn.execute(
                """UPDATE idempotency_keys SET
                       status        = 'completed',
                       status_code   = $4,
                       response_body = $5,
                       request_hash  = $6,
                       expires_at    = NOW() + ($7 || ' seconds')::interval
                     WHERE tenant_id = $1::uuid AND key = $2 AND scope = $3""",
                tenant_id, idem_key, scope,
                response.status_code, parsed_body, hash_, str(COMPLETED_TTL_SEC),
            )
        try:
            await as_system(_complete)
        except Exception:
            pass

        return Response(
            content=captured,
            status_code=response.status_code,
            headers=dict(response.headers),
            media_type=response.media_type,
        )
